### DISCUSSION TITLE
Proposal: The boundaries of the shape no longer include the border.

### DISCUSSION THREAD

**Discussion by zhe-he** [fabricjs/fabric.js#9805] [Ideas]
1. Visual Consistency: In design and layout, it may be necessary for the actual occupied space of a shape to remain consistent with its visual representation. Excluding the border from the boundary calculation can ensure more precise layout positioning of the shape, avoiding layout shifts caused by changes in border width.
﻿
2. Interaction Design: In interaction design, the border may serve only as a visual decoration and should not affect the actual interactive area of the shape. For example, the area responsive to mouse clicks or touches should be based on the main body of the shape, not its decorative border.
﻿

The consequences: After setting the border size, the effect of the old version selected on the left will all turn into the effect shown on the right.

<img width="244" alt="image" src="https://github.com/fabricjs/fabric.js/assets/11306583/9d5af6c2-4970-42b8-8361-862b9a0559f5">
<img width="232" alt="image" src="https://github.com/fabricjs/fabric.js/assets/11306583/88fdafea-1435-456e-9a88-f14e177ebd6e">

**Comment by ShaMan123** [comment]
It is planned to remove originX/Y thus positioning the object by its center internally while keeping the methods that are able to position objects.
Those methods could be overridden to accommodate any needs and could as well include flags such as you suggest here.
I suggest you work only with originX/Y `center` from now on.
#9736

**Comment by asturur** [comment]
Fabric around 1.3 or 1.4 had borders across the bounding box, but then after many requests it was moved as it is today.
The issue with the 'border is not part of the shape' apart the debate itself is that for hollow shapes the border is the body.
If i give you an empty rect that is made only with a large stroke or a path that is rectangular and hollow inside, the 2 objects will behave differently, one will have the bounding box over the stroke, the other outside. But conceptually those objects are the same.

I think that whoever wants to have fabric behave in that way can probably find a path either in the render logic or in the geometry functions to exclude the stroke.
Like setting the stroke to 0 and painting something other new property during renderStroke

**Comment by zhe-he** [comment]
@asturur 
Well, it's like this. I have a requirement similar to Figma, where I need to set the shape's outer border, inner border, and the default middle border. If the shape's boundaries are not defined to exclude the borders first, then the above requirements are difficult to achieve in a reasonable manner, and it would result in writing a lot of cumbersome code, which is quite painful. If no modifications are made, then how should this kind of requirement for different types of borders be updated on the existing basis?

In [simple demo](https://codesandbox.io/p/devbox/fabric-stroke-side-x6s4pz?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clptda5ls00073b6hub7q0rb8%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clptda5lo00023b6h7o2lx4gg%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clptda5lr00043b6h1ko80kdv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clptda5ls00063b6hyx3tfc52%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clptda5lo00023b6h7o2lx4gg%2522%253A%257B%2522id%2522%253A%2522clptda5lo00023b6h7o2lx4gg%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clptda5ls00063b6hyx3tfc52%2522%253A%257B%2522id%2522%253A%2522clptda5ls00063b6hyx3tfc52%2522%252C%2522activeTabId%2522%253A%2522clto50vgc00ea3b6gqczq5drm%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clto50vgc00ea3b6gqczq5drm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clptda5lr00043b6h1ko80kdv%2522%253A%257B%2522id%2522%253A%2522clptda5lr00043b6h1ko80kdv%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clptda5lr00033b6hp9wzt2bi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clptda5lr00033b6hp9wzt2bi%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D), I only wrote rect2 to implement the requirements, but there was a bug like the one mentioned in the comment about the border occupying position. 
In [Complete demo](https://codesandbox.io/p/devbox/fabric-stroke-patch-nnyrlw?file=%2Fsrc%2Ftest-stroke-side.tsx%3A29%2C17), I rewrote the logic of calculating boundaries in the graph to fix this bug.
[#9727](https://github.com/fabricjs/fabric.js/discussions/9727)

**Reply by asturur** [reply]
This idea works only with simple geometric shapes and when you move to thing like Path or Text requires multiple composite operation to be done.
This feature also does not work with padding or make padding unclear at best.
The way to fix this without too much cumbersome code, is to set always stroke:0 to your code, so that stroke is excluded and build your new property to draw inner/middle/outer border.

**Reply by vaynevayne** [reply]
and build your new property to draw inner/middle/outer border 
Hello. What does the above statement mean? Could you provide some basic code? Thanks

**Reply by asturur** [reply]
It is a lot of code, especially for outer and inner border that need drawing on a different canvas and then compose together

**Reply by vaynevayne** [reply]
I did not understand, my scene is very simple, just draw a rectangle, how do I draw innerBorder?

**Reply by zhe-he** [reply]
> I did not understand, my scene is very simple, just draw a rectangle, how do I draw innerBorder?

这是一个很大的工程，涉及到的地方包括但不限于：
1 缩放矩形，矩形的边框可以缩放或不缩放
2 边框可以设置圆角，如果是外边框或者内边框，它的圆角应该怎么计算
3 边框可以设置渐变，外边框起始点的定义影响到的缓存怎么处理
4 如果给边框填充图片，单独水平缩放，背景图片应该怎么变更
5 路径对象，他的边框应该怎么定义

你的需求简单的处理方式为：
把边框设置为0，然后你新增2个属性自定义边框，自定义描边类型。然后重写render

This is a large project, involving but not limited to:
1. Scaling rectangles, where the rectangle's border can be scaled or not.
2. Borders can have rounded corners; how should the rounding be calculated for outer or inner borders?
3. Borders can have gradients; how to handle the cache affected by the definition of the starting point of the outer border?
4. If an image is filled in the border, how should the background image change with separate horizontal scaling?
Path objects; how should their borders be defined?

A simple way to handle your requirements is:
Set the border to 0, then add 2 new properties to customize the border and the custom stroke type. Then override the render method.

**Reply by asturur** [reply]
To draw an innerborder you have to draw your rectangle in isolation and use a global composition operation of 'source-in' between the fill and the stroke, that is something that fabricJS does not do by default. So you have to create your custom classes

**Reply by vaynevayne** [reply]
How to create a custom class? Can you provide a little demo?

**Reply by vaynevayne** [reply]
方便加微信吗? @zhe-he

**Comment by usamshen** [comment]
我建了一个微信群 我们可以将用fabric的使用中文的人拉在一起互相帮助 方便一点

**Comment by zhe-he** [comment]
@vaynevayne 
提议中的2个例子，就是你要的自定义图形

The two examples proposed are the custom graphics you want.
